usr/sbin/p11sak/p11sak.c

```diff
@@ -1908,106 +2053,277 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
                 printf(" %*s | ", CELL_SIZE, keytype);
             printf("%s\n", label);
         }
+        free(label);
+        free(keytype);
     }
 
-    ret = funcs->C_FindObjectsFinal(session);
-    if (ret != CKR_OK) {
-        printf("C_FindObjectsFinal failed (error code 0x%lX)\n", ret);
+    rc = funcs->C_FindObjectsFinal(session);
+    if (rc != CKR_OK) {
+        printf("C_FindObjectsFinal failed (error code 0x%lX: %s)\n", rc,
+                p11_get_ckr(rc));
         goto done;
     }
 
-    ret = CKR_OK;
+    rc = CKR_OK;
+
+done:
+
+    return rc;
+}
+
+static CK_RV getKeyboardInput(char *buff, char *label)
+{
+
+    CK_RV rc = CKR_OK;
+    int nread;
+    size_t buflen;
+
+    printf("Are you sure to delete %s? (y/n)\n", label);
+
+    nread = getline(&buff, &buflen, stdin);
+    if (nread == -1) {
+        rc = -1;
+    }
+    buff[nread - 1] = '\0';
+
+    return rc;
+}
+
+static CK_RV finalizeDestroyObject(char *buff, char *label, CK_SESSION_HANDLE *session,
+                                   CK_OBJECT_HANDLE *hkey)
+{
+    CK_RV rc = CKR_OK;
+
+    if (strcmp(buff, "y") == 0) {
+        printf("Destroy Object with Label: %s\n", label);
+        rc = funcs->C_DestroyObject(*session, *hkey);
+        if (rc != CKR_OK) {
+            printf(
+                    "Key with label %s could not be destroyed (error code 0x%lX: %s)\n",
+                    label, rc, p11_get_ckr(rc));
+            return rc;
+        }
+        printf("DONE - Destroy Object with Label: %s\n", label);
+    } else if (strcmp(buff, "n") == 0) {
+        printf("Skip deleting Key\n");
+    } else {
+        printf("Please just enter (y) for yes or (n) for no.\n");
+    }
 
-    done: free(label);
-    free(keytype);
-    return ret;
+    return rc;
 }
 
-static CK_RV execute_cmd(CK_SESSION_HANDLE session, CK_SLOT_ID slot, p11sak_cmd cmd,
-        p11sak_kt kt, CK_ULONG keylength, CK_ULONG exponent, char *ECcurve,
-        char *label, char *attr_string, int long_print) {
-    CK_RV ret;
+/**
+ * Delete objects
+ */
+static CK_RV deleteKey(CK_SESSION_HANDLE session, p11sak_kt kt, char *rm_label,
+                       CK_BBOOL *forceAll)
+{
+    CK_ULONG keylength, count;
+    CK_OBJECT_CLASS keyclass;
+    CK_OBJECT_HANDLE hkey;
+    char *keytype = NULL;
+    char *label = NULL;
+    CK_RV rc;
+    char *buff = NULL;
```

Comment:
I would not call this 'buff', name it 'user_input', or anything that tells the reader what it is used for.